basic_features/basic_mod_data_checker.py

```diff
@@ -0,0 +1,258 @@
+
+        Returns:
+            The 0-based index of the matched group or None for no match.
+        """
+        if (pattern := self.get(key)) is None:
+            return None
+        else:
+            return self.match_index_of_pattern(search_str, pattern)
+
+    @staticmethod
+    def match_index_of_pattern(search_str: str, pattern: re.Pattern) -> Optional[int]:
+        """Get the index of the matched group if `search_str` matches `pattern`
+        (from this dict).
+
+        Returns:
+            The 0-based index of the matched group or None for no match.
+        """
+        if not (match := pattern.match(search_str)) or match.lastindex is None:
+            return None
+        return match.lastindex - 1
+
+    @staticmethod
+    def regex_from_glob_list(glob_list: Iterable[str]) -> re.Pattern:
+        """Returns a regex pattern form a list of glob patterns.
+
+        Every pattern has a capturing group, so that `match.lastindex - 1` will
+        give the `glob_list` index.
+        """
+        return re.compile(
+            f'(?:{"|".join(f"({fnmatch.translate(f)})" for f in glob_list)})', re.I
+        )
+
+
+class GlobPatternDict(TypedDict, total=False):
+    """See: `BasicModDataChecker`"""
+
+    unfold: Iterable[str] | None
+    valid: Iterable[str] | None
+    delete: Iterable[str] | None
+    move: MutableMapping[str, str] | None
+
+
+class BasicModDataChecker(mobase.ModDataChecker):
+    """Game feature that is used to check and fix the content of a data tree
+    via simple file definitions.
+
+    The file definitions support glob pattern (without subfolders) and are checked in
+    definition order (passed either as dict or kwargs).
+
+    Args:
+        file_patterns (optional): An dict with the keys below or as kwargs:
+
+        unfold (optional): Folders to unfold (remove and move contents to parent),
+            after being checked and fixed recursively.
+
+        valid (optional): Files and folders in the right path.
+            Check result: `mobase.ModDataChecker.VALID`.
+
+        delete (optional): Files/folders to delete.
+            Check result: `mobase.ModDataChecker.FIXABLE`.
+
+        move (optional): Files/folders to move and their target.
+            If the path ends with `/` or `\\`, the entry will be inserted
+            in the corresponding directory instead of replacing it.
+
+            Check result: `mobase.ModDataChecker.FIXABLE`.
+
+        Example::
+
+            BasicModDataChecker(
+                valid=["valid_folder", "*.ext1"]
+                move={"*.ext2": "path/to/target_folder/"}
+            )
+
+            BasicModDataChecker(
+                {
+                    "valid": ["valid_folder", "*.ext1"],
+                    "move": {"*.ext2": "path/to/target_folder/"}
+                }
+            )
+
+    See Also:
+        `mobase.IFileTree.move`
+    """
+
+    file_patterns = None  # type: GlobPatternDict
```

Comment:
Why do you use a `# type` annotation here an standard typing annotation everywhere else?

Reply:
Because you cannot assign None initially to a strictly typed (not optional) field, with the new typing annotation, which you have to initialize in the constructor, since it is mutable. That is why you have a `default_factory` in `dataclasses`. The only other option would be to define the instance var in the constructor only.

This is a compromise allow default `file_patterns` to be defined in a derived class as a field, as well as args to the constructor., while ensuring the mutable field is not shared between instances.

If you prefer another solution, like strictly instance only vars, I can move it into the constructor.

Reply:
See my comment on the `__init__`: either use an instance only vars that you pass to `super().__init__` in the child classes or use a different name for the class var (if I understood the intent correctly).

Reply:
Changed in f3977a9: `default_file_patterns: ClassVar`, the rest instance vars.